server/src/com/thoughtworks/go/server/messaging/plugin/StageStatusPluginNotifier.java

```diff
@@ -16,49 +16,145 @@
-        Map data = createRequestDataMap(stage);
+        if (notificationPluginRegistry.isAnyPluginInterestedIn(NotificationExtension.STAGE_STATUS_CHANGE_NOTIFICATION)) {
+            Map data = createRequestDataMap(stage);
 
-        pluginNotificationQueue.post(new PluginNotificationMessage(NotificationExtension.STAGE_STATUS_CHANGE_NOTIFICATION, data));
+            pluginNotificationQueue.post(new PluginNotificationMessage(NotificationExtension.STAGE_STATUS_CHANGE_NOTIFICATION, data));
+        }
     }
 
     Map createRequestDataMap(Stage stage) {
-        Map data = new LinkedHashMap();
-        data.put("pipeline-name", stage.getIdentifier().getPipelineName());
-        data.put("pipeline-counter", stage.getIdentifier().getPipelineCounter());
-        data.put("stage-name", stage.getIdentifier().getStageName());
-        data.put("stage-counter", stage.getIdentifier().getStageCounter());
-        data.put("stage-state", stage.getState());
-        data.put("stage-result", stage.getResult());
-        data.put("create-time", timestampToString(stage.getCreatedTime()));
-        data.put("last-transition-time", timestampToString(stage.getLastTransitionedTime()));
+        Map<String, Object> data = new LinkedHashMap<String, Object>();
+        String pipelineName = stage.getIdentifier().getPipelineName();
+        Integer pipelineCounter = new Integer(stage.getIdentifier().getPipelineCounter());
+
+        Map<String, Object> pipelineMap = createPipelineDataMap(pipelineName, pipelineCounter, stage);
+        data.put("pipeline", pipelineMap);
+
         return data;
     }
 
+    private Map<String, Object> createPipelineDataMap(String pipelineName, Integer pipelineCounter, Stage stage) {
+        Map<String, Object> pipelineMap = new LinkedHashMap<String, Object>();
+        pipelineMap.put("name", pipelineName);
+        pipelineMap.put("counter", pipelineCounter.toString());
+
+        String pipelineGroup = goConfigService.findGroupNameByPipeline(new CaseInsensitiveString(pipelineName));
+        pipelineMap.put("group", pipelineGroup);
+
+        BuildCause buildCause = pipelineSqlMapDao.findBuildCauseOfPipelineByNameAndCounter(pipelineName, pipelineCounter);
+
+        List<Map> materialRevisionList = createBuildCauseDataMap(buildCause.getMaterialRevisions());
+        pipelineMap.put("build-cause", materialRevisionList);
+
+        Map<String, Object> stageMap = createStageDataMap(stage);
+        pipelineMap.put("stage", stageMap);
+
+        return pipelineMap;
+    }
+
+    private List<Map> createBuildCauseDataMap(MaterialRevisions materialRevisions) {
+        List<Map> materialRevisionList = new ArrayList<Map>();
+        for (MaterialRevision currentRevision : materialRevisions) {
+            Map<String, Object> materialRevisionMap = new LinkedHashMap<String, Object>();
+
+            materialRevisionMap.put("material", ((InformationProvider) currentRevision.getMaterial()).getAttributes(false));
```

Comment:
Oh, come on, a small change like this make the material a proper ```InformationProvider``` and doesn't need casts like this all over. You implemented the polymorphism change all over. Why wouldn't you just do this? It could even be marked abstract, since all the descendants of ```Material``` already implement it. Doing it this way just leads to someone seeing the ```Material``` interface, not implementing this method, and it'll fail with a ```ClassCastException``` at runtime, somewhere.

```
diff --git a/common/src/com/thoughtworks/go/config/materials/AbstractMaterial.java b/common/src/com/thoughtworks/go/config/materials/AbstractMaterial.java
index 9fe8fac..bfaef5c 100644
--- a/common/src/com/thoughtworks/go/config/materials/AbstractMaterial.java
+++ b/common/src/com/thoughtworks/go/config/materials/AbstractMaterial.java
@@ -182,4 +182,9 @@ public abstract class AbstractMaterial extends PersistentObject implements Mater
     public MaterialConfig config() {
         throw new RuntimeException("You need to implement this");
     }
+
+    @Override
+    public Map<String, Object> getAttributes(boolean addSecureFields) {
+        return new HashMap<String, Object>();
+    }
 }
diff --git a/common/src/com/thoughtworks/go/domain/materials/Material.java b/common/src/com/thoughtworks/go/domain/materials/Material.java
index 9138ddd..7c21fa5 100644
--- a/common/src/com/thoughtworks/go/domain/materials/Material.java
+++ b/common/src/com/thoughtworks/go/domain/materials/Material.java
@@ -16,10 +16,6 @@
 
 package com.thoughtworks.go.domain.materials;
 
-import java.io.File;
-import java.io.Serializable;
-import java.util.Map;
-
 import com.thoughtworks.go.config.CaseInsensitiveString;
 import com.thoughtworks.go.config.PipelineConfig;
 import com.thoughtworks.go.config.materials.SubprocessExecutionContext;
@@ -29,7 +25,11 @@ import com.thoughtworks.go.util.command.EnvironmentVariableContext;
 import com.thoughtworks.go.util.command.ProcessOutputStreamConsumer;
 import com.thoughtworks.go.util.json.JsonMap;
 
-public interface Material extends Serializable {
+import java.io.File;
+import java.io.Serializable;
+import java.util.Map;
+
+public interface Material extends Serializable, InformationProvider {
 
     //-- DB and config behaviour

```